Fix indexing boolean ranges with empty ranges

jishnub opened the pull request:
Fixes #40760 

Now
```julia
julia> (false:false)[1:0]
true:false

julia> (false:true:false)[1:1:0]
0:1:-1

julia> (false:true:false)[1:0]
0:1:-1
```

jishnub committed:
indexing ranges with empty ranges

jishnub commented:
bump

ViralBShah commented:
@mbauman Any thoughts here? If you generally think this is ok, we can get it rebased and merged.

jishnub committed:
fix inference

jishnub committed:
fix comment

ViralBShah commented:
I believe this is good to merge.

DilumAluthge commented:
@ViralBShah @mbauman CI is green, but it's not clear whether or not this PR has had an approving review.

N5N3 commented on base/range.jl:
```diff
@@ -1004,19 +1008,9 @@ function getindex(r::StepRange, s::AbstractRange{T}) where {T<:Integer}
-            start, len = last(r), 1
-        end
-        return range(start, step=step(r); length=len)
+        range(first(s) ? first(r) : last(r), step = step(r), length = Int(last(s)))
```
This branch seems broken if `s` isa `StepRangeLen{Bool}`
```julia
julia> (1:2:3)[StepRangeLen{Bool}(true,-1,2)]
1:2:0  #3:2:3 on master (also not correct but at least not empty.)

julia> (1:2:3)[StepRangeLen{Bool}(true,-1,2) |> collect]
1-element Vector{Int64}:
 1
```
Since this change doesn't help to fix #40760, should we keep the version on master?

vtjnash commented on base/range.jl:
```diff
@@ -988,6 +988,9 @@ function getindex(r::AbstractUnitRange, s::AbstractUnitRange{T}) where {T<:Integ
+        isempty(s) && return range(oneunit(f), zero(f))
         start = oftype(f, f + first(s) - firstindex(r))
         len = length(s)
         stop = oftype(f, start + (len - oneunit(len)))
```
Or simply `start, stop = stop, start` in that branch, so we don't need to attempt to do any arithmetic, but simply preserve the values given

vtjnash committed:
fixup! fix comment

vtjnash committed:
squash! indexing ranges with empty ranges

Reviewing the work here, it seems undesirable to me now to actually to
make the majority of this change. We already define too much of what the
value of start should be, and trying to assume that one(T):zero(T) is in
range for T<:Integer is not necessarily ideal either.

This keeps some minor semi-related changes though:
 - ensured this only happens when first(r)+first(s)*st == typemax
 - if the range is supposed to be empty, then use a step of eps=1 (as
   steprange_last_empty will do later), to reduce the size of the set of
   values that overflow and become !isempty or error to 1 value instead of
   step values
 - Handle a convoluted case of StepRangeLen{Bool} indexing into a Vector
   returning the element from the wrong end

```
julia> (typemin(Int)+1:100:zero(Int))[1:0] |> length
184467440737095516 # old
0 # new
```

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/40764#issuecomment-1793011629) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b64b578_vs_4462257/report.md).
